cloud-storage-uploader/src/upload-helper.ts

```diff
@@ -0,0 +1,91 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Storage, UploadResponse } from '@google-cloud/storage';
+import * as path from 'path';
+import { getFiles } from './util';
+
+export class UploadHelper {
+  readonly storage: Storage;
+  constructor(storage: Storage) {
+    this.storage = storage;
+  }
+  /**
+   * Uploads a file to a bucket
+   * based on https://github.com/googleapis/nodejs-storage/blob/master/samples/uploadFile.js
+   * @param bucketName The name of the bucket
+   * @param filename The file path
+   * @param destination The destination prefix
+   */
+  async uploadFile(
+    bucketName: string,
+    filename: string,
+    destination?: string,
+  ): Promise<UploadResponse> {
+    interface UploadOptions {
+      gzip: boolean;
+      destination?: string;
+    }
+    const options: UploadOptions = { gzip: true };
+    if (destination) {
+      // if obj prefix is set, then extract filename and append to prefix
+      options.destination = `${destination}/${path.posix.basename(filename)}`;
+    }
+    const uploadedFile = await this.storage
+      .bucket(bucketName)
+      .upload(filename, options);
+    return uploadedFile;
+  }
+
+  /**
+   * Uploads a specified directory to a GCS bucket
+   * based on https://github.com/googleapis/nodejs-storage/blob/master/samples/uploadDirectory.js
+   * @param bucketName The name of the bucket
+   * @param directoryPath The path of the directory to upload
+   * @param objectKeyPrefix Optional Prefix for in the GCS bucket
+   * @param clearExistingFilesFirst Clean files in the prefix before uploading
+   */
+  async uploadDirectory(
```

Comment:
This is probably fine for v1, but we should take a follow-up to limit the number of concurrent uploads here. Right now, if there are 20,000 files, this is going to create 20,000 async jobs and that could make for some sadness

Reply:
@sethvargo I would be happy to take shot at this now, some googling led me to https://www.npmjs.com/package/bottleneck and https://www.npmjs.com/package/es6-promise-pool which are lightweight and has no deps. Thoughts @sethvargo?